modules/user/lib/fantom/user-sync-reliquary-farm-balance.service.ts

```diff
@@ -0,0 +1,264 @@
+import { networkConfig } from '../../../config/network-config';
+import { reliquaryService } from '../../../subgraphs/reliquary-subgraph/reliquary.service';
+import ReliquaryAbi from '../../../web3/abi/Reliquary.json';
+import { getContractAt, jsonRpcProvider } from '../../../web3/contract';
+import { Multicaller } from '../../../web3/multicaller';
+import { Reliquary } from '../../../web3/types/Reliquary';
+import { UserStakedBalanceService, UserSyncUserBalanceInput } from '../../user-types';
+
+type ReliquaryPosition = {
+    amount: BigNumber;
+    rewardDebt: BigNumber;
+    rewardCredit: BigNumber;
+    entry: BigNumber;
+    poolId: BigNumber;
+    level: BigNumber;
+    genesis: BigNumber;
+    lastMaturityBonus: BigNumber;
+};
+
+type BalanceChangedEvent = Event & {
+    args: {
+        pid: BigNumber;
+        amount: BigNumber;
+        to: string;
+        relicId: BigNumber;
+    };
+};
+
+type RelicManagementEvent = Event & {
+    args: {
+        fromId: BigNumber;
+        toId: string;
+        amount: BigNumber;
+    };
+};
+
+type TransferEvent = Event & {
+    args: {
+        from: string;
+        to: string;
+        tokenId: BigNumber;
+    };
+};
+
+export class UserSyncReliquaryFarmBalanceService implements UserStakedBalanceService {
+    constructor() {}
+
+    public async syncChangedStakedBalances(): Promise<void> {
+        const status = await prisma.prismaUserBalanceSyncStatus.findUnique({ where: { type: 'RELIQUARY' } });
+
+        if (!status) {
+            throw new Error('UserReliquaryFarmBalanceService: syncStakedBalances called before initStakedBalances');
+        }
+
+        const pools = await prisma.prismaPool.findMany({ include: { staking: true } });
+        const latestBlock = await jsonRpcProvider.getBlockNumber();
+        const farms = await reliquaryService.getAllFarms({});
+
+        const startBlock = status.blockNumber + 1;
+        const endBlock = latestBlock - startBlock > 10_000 ? startBlock + 10_000 : latestBlock;
+        const amountUpdates = await this.getAmountsForUsersWithBalanceChangesSinceStartBlock(
+            networkConfig.masterchef.address,
+            startBlock,
+            endBlock,
+        );
+        const userAddresses = _.uniq(amountUpdates.map((update) => update.userAddress));
+
+        if (amountUpdates.length === 0) {
+            await prisma.prismaUserBalanceSyncStatus.update({
+                where: { type: 'RELIQUARY' },
+                data: { blockNumber: endBlock },
+            });
+
+            return;
+        }
+
+        await prismaBulkExecuteOperations(
+            [
+                prisma.prismaUser.createMany({
+                    data: userAddresses.map((userAddress) => ({ address: userAddress })),
+                    skipDuplicates: true,
+                }),
+                ...amountUpdates.map((update) => {
+                    const pool = pools.find((pool) => pool.staking?.id === update.farmId);
```

Comment:
that could also match on a mastechef farm id?
need to only get pools with reliquary staking type

Reply:
true, good catch!

Reply:
also added this in the master chef version